### PR TITLE
Clear workflow contexts on History Shard closure

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

Clear all workflow contexts, found in the cache, that belong to a History Shard that is being closed. 

## Why?

Right now, when a History Shard is closed, any non-durable Update that is still in-flight will be "lost", meaning that there is no way to access it, cancel it or complete it anymore. The client that issued the Update is not informed about this either (neither is the Frontend). 

A better experience is to forcibly cancel any in-flight Updates belonging to the History Shard that is closing. That way, the Frontend can re-issue the Update request (which will land on a new History Shard). The client won't even know that the History Shard was closed. However, if the retry happens towards the end of the Update timeout, the client's request might still time out after all.

## How did you test it?

Added unit tests and functional test.

Before going live, this will have to be validated by thorough end-to-end testing.

### Code Complexity

Since the workflow cache already has all the workflow contexts, I had considered adding a by-shardID index a method to iterate over them there. However, its cache implementation is based on `cache.Cache` (ie a generic cache) it would have been awkward to add it there. Therefore, I decided to track the workflow contexts in a separate place, the Finalizer, which adds to the complexity.

### Memory Leak

There is no way to list the workflow contexts that belong to a History Shard. Therefore, a new "Finalizer" was introduced that maintains a list of cleanup callbacks: one callback per workflow context, grouped by History Shard.

But since the workflow cache will eventually evict workflow contexts that are no longer needed, if the Finalizer holds on to these cleanup callbacks past the lifetime of the workflow context, the GC won't be able to free them from memory. Therefore, a lifecycle hook was introduced that removes any expired workflow contexts from the Finalizer.

### Delayed History Shard Closure

The step to close the History Shard should not lock up if some/any of the workflow contexts cannot be cleared. To clear them, the lock for the workflow needs to be obtained. To prevent this from blocking indefinitely, a timeout was introduced. It's better to make sure the History Shard closes eventually than to clear all workflow contexts.

PS: Logging has been added to be able to monitor the timeouts and completion rate of the Finalizer.

PPS: If any production issues should occur, the Finalizer's cleanup step can be disabled entirely by setting the dynamic config for the timeout to zero. However, the registering/de-registering of callbacks cannot be disabled.

### CODE REVIEW

File: common/cache/cache.go
Comment (stephanos) [inline common/cache/cache.go:114]: These two callbacks are introduced to allow callbacks to fire based on the cache item's lifecycle events.

File: common/cache/lru.go
Comment (stephanos) [inline common/cache/lru.go:338]: This code might need a refactor that's out of scope of this PR. Therefore, this code needed to be added twice.

File: common/finalizer.go
Comment (stephanos) [inline common/finalizer.go:124]: I decided to use the adaptive pool here to allow for auto-scaling of the goroutines needed to complete the job.

File: common/finalizer_test.go
Comment (stephanos) [inline common/finalizer_test.go:134]: NOTE: I decided to not re-test the adaptive pool behavior here again. This leaves a bit of a blind spot, but re-testing it seemed over the top.

File: common/finalizer.go
Comment (stephanos) [inline common/finalizer.go:25]: The Finalizer is generic and knows nothing about shards of mutable contexts, so I decided to put it into the common package.

File: service/history/shard/context_impl.go
Comment (stephanos) [inline service/history/shard/context_impl.go:1482]: This appears to be the best place to invoke the finalizer.

File: service/history/workflow/cache/cache.go
Comment (stephanos) [inline service/history/workflow/cache/cache.go:97]: The cache value is wrapped with this new struct to provide access to the finalizer etc. It is hidden from the caller.

File: service/history/workflow/cache/cache.go
Comment (stephanos) [inline service/history/workflow/cache/cache.go:461]: There is not much we can do if any of these operations fail, other than logging it.

File: common/cache/lru.go
Comment (stephanos) [inline common/cache/lru.go:313]: I believe this is an oversight; it should use the timesource and not `time` directly. Without this, my test wasn't passing.

File: common/dynamicconfig/constants.go
Comment (stephanos) [inline common/dynamicconfig/constants.go:1333]: Q: Is this a reasonable timeout?

File: tests/update_workflow.go
Comment (stephanos) [inline tests/update_workflow.go:3545]: This and other tests are asserting the behavior of a "lost Update" (ie one that isn't actively aborted on shard movements). Therefore, the functionality is disabled for these.

File: service/history/shard/context_impl.go
Comment (stephanos) [inline service/history/shard/context_impl.go:1486]: This appears to be the best place to invoke the finalizer.

File: common/finalizer.go
Comment (stephanos) [inline common/finalizer.go:99]: I decided to use the existing adaptive pool here to allow for auto-scaling of the goroutines needed to complete the job.

File: common/finalizer.go
Comment (dnr) [reply common/finalizer.go:25]: Please use another package, or a new package just for this. `common` is way-overloaded and is the source of a lot of dependency cycles, I've been moving stuff out of it slowly

Comment (stephanos) [reply common/finalizer.go:25]: I can do that 👍

File: common/finalizer/finalizer_test.go
Comment (stephanos) [inline common/finalizer/finalizer_test.go:140]: NOTE: I decided to not re-test the adaptive pool behavior here again. This leaves a bit of a blind spot, but re-testing it seemed over the top.

File: service/history/workflow/cache/cache.go
Comment (stephanos) [inline service/history/workflow/cache/cache.go:180]: There is not much we can do if any of these operations fail, other than logging it.

File: common/finalizer/finalizer.go
Comment (alexshtin) [inline common/finalizer/finalizer.go:108]:
```suggestion
		f.logger.Warn("finalizer skipped: zero timeout")
```

File: common/finalizer/finalizer.go
Comment (alexshtin) [inline common/finalizer/finalizer.go:128]: Why do you need this func and another go routine? `pool.Do` is async, you should be able just call it for every `cb`.

File: tests/update_workflow.go
Comment (alexshtin) [inline tests/update_workflow.go:184]: Can you add a comment here, explaining why `DescribeMutableState` is used for this. Similar to ones I had before:
```
// Side effect of DescribeMutableState is clear of workflow context because
// it needs to return both cached and database versions.
```

**Review by alexshtin — APPROVED** [review]

File: common/finalizer/finalizer.go
Comment (stephanos) [reply common/finalizer/finalizer.go:128]: Ah, you're right. That's still left from before I added the pool. I'll remove it.

Comment (stephanos) [reply common/finalizer/finalizer.go:128]: Actually ... looking closer, wouldn't this trigger the for loop scoping issue if I remove it? https://go.dev/blog/loopvar-preview

Comment (alexshtin) [reply common/finalizer/finalizer.go:128]: In go 1.22, it is on by default. So it should capture right `cb` on every iteration.

File: common/dynamicconfig/constants.go
Comment (alexshtin) [reply common/dynamicconfig/constants.go:1346]: How many updates can be aborted within 2 seconds? I bet quite a lot. Let's start with 2 seconds and see. I don't have better number.

File: common/finalizer/finalizer.go
Comment (stephanos) [reply common/finalizer/finalizer.go:128]: Ah! I didn't catch that it was already enabled now. I thought it was still in preview. Time flies.

File: common/finalizer/finalizer.go
Comment (stephanos) [inline common/finalizer/finalizer.go]: @alexshtin I noticed there was a potential goroutine leak here: if the finalizer times out, this goroutine would have never been completed since the waitgroup would have never been done. So I'm counting the completed items myself now while also listening for the timeout to abandon the goroutine, if needed.

File: common/dynamicconfig/constants.go
Comment (stephanos) [reply common/dynamicconfig/constants.go:1346]: No clue, really. Let's start there 👍

File: service/history/ndc/activity_state_replicator_test.go
Comment (yycptt) [inline service/history/ndc/activity_state_replicator_test.go:98]: Curious what's special about this test suite that requires cleanup logic to be disabled? Shall we disable it by default and only enable for tests that are specifically testing the clean up?

File: common/finalizer/finalizer.go
Comment (yycptt) [inline common/finalizer/finalizer.go:73]: there could be a race condition here?

File: common/finalizer/finalizer.go
Comment (yycptt) [inline common/finalizer/finalizer.go:154]: why this needs to be in a separate goroutine?

File: common/finalizer/finalizer.go
Comment (yycptt) [inline common/finalizer/finalizer.go:137]: this goroutine will continue to make callback even after the timeout?

File: service/history/ndc/activity_state_replicator_test.go
Comment (stephanos) [reply service/history/ndc/activity_state_replicator_test.go:98]: Right; so this test suite (and the other one as well) will invoke `s.mockShard.StopForTest()` which eventually triggers the Finalizer to run. Which in and of itself is absolutely no problem. However, these test suites use a `workflow.MockContext` with a couple of `EXPECT`s. I just don't see any value in mocking the Lock, Unlock and Clear calls for every single test case. So I disabled them explicitly for these 2 tests suites.

Now, having said all this ... I think the better approach is to just remove the Finalizer from the mock shard entirely. So I'll do that next.

File: common/finalizer/finalizer.go
Comment (stephanos) [reply common/finalizer/finalizer.go:137]: So the way the `pool.Do` works is that it will block once there are not enough workers to handle the work. But also, once you call `Stop`, all further `Do` calls become a noop. It will still finish in-flight work, but after that, it will not execute any of the remaining callbacks.

From its documentation:
```
// Do calls f() on a worker goroutine. If the call can't be started within targetDelay, it adds
// another worker. If Stop is called concurrently, Do may or may not call f. If Stop has been
// called already, Do does nothing. 
```

File: common/finalizer/finalizer.go
Comment (stephanos) [reply common/finalizer/finalizer.go:154]: I think you're right. This was using a WaitGroup before, but now that I've removed that, I can merge this select with the next select. I'll do that!

File: common/finalizer/finalizer.go
Comment (stephanos) [reply common/finalizer/finalizer.go:73]: Oh, I see it now! The for loop can still iterate over the callbacks after the mutex was unlocked. I'm going to add a step to ensure that the goroutine completes before returning and releasing the lock. Since at that point, the pool has stopped, the calls to `Do` won't take a lot of time anymore. So I think it's safe to wait for that.

File: common/finalizer/finalizer.go
Comment (yycptt) [inline common/finalizer/finalizer.go:162]: nit: I'd probably always do this maybe in a defer or something. I understand in completed case it will always be unblocked, but that took me some time to realize that when reading the code. No strong opinion.

File: common/finalizer/finalizer.go
Comment (yycptt) [reply common/finalizer/finalizer.go:73]: oh sorry, I was actually referring to something else. I am thinking there's a race condition between Register() and Run() methods. Register() can first pass the finalized check, Run() method then change the flag, complete all the callbacks, then Register adds the new callback into the callbacks list but it will never be invoked.


But thanks for catching the other issue!

File: common/finalizer/finalizer.go
Comment (stephanos) [reply common/finalizer/finalizer.go]: Good feedback; I've added a defer!

File: common/finalizer/finalizer.go
Comment (stephanos) [reply common/finalizer/finalizer.go]: That's true. And in terms of logic that's fine, once the Finalizer is started, any callbacks that are added after it completes will be ignored. But this prompted me to re-visit this since the call to Register would have to block until the Run completes. Which is not ideal. So I now re-structured it a bit to ensure that Register/Deregister unblock immediately when Run is/was running.

File: common/finalizer/finalizer.go
Comment (stephanos) [inline common/finalizer/finalizer.go]: This is not needed anymore since Register/Deregister will not access or modify callbacks anymore once Run starts the goroutine.

**Review by yycptt — APPROVED** [review]

File: common/finalizer/finalizer.go
Comment (yycptt) [reply common/finalizer/finalizer.go]: 👍 yeah I mainly want the finalizer to have a clear contract with the caller and caller can know for sure if the callback registered will be eventually invoked or not.

**Pull request merged**
